hhderiv.mod: Use of celsius variable with derivmplicit solver leads to error "Static variables are not supported in acc routine"

Consider hhderiv.mod file from testcorenrn test suite: https://github.com/neuronsimulator/testcorenrn/blob/master/mod/hhderiv.mod

This has:

```python

? states
DERIVATIVE states {
        rates(v)
        m' =  (minf-m)/mtau
        h' = (hinf-h)/htau
        n' = (ninf-n)/ntau
}

:LOCAL q10


? rates
PROCEDURE rates(v(mV)) {  :Computes rate and other constants at current v.
                      :Call once from HOC to initialize inf at resting v.
        LOCAL  alpha, beta, sum, q10
        :TABLE minf, mtau, hinf, htau, ninf, ntau DEPEND celsius FROM -100 TO 100 WITH 200

UNITSOFF
        q10 = 3^((celsius - 6.3)/10)
```

and this result into code like:

```cpp

    int states_hhderiv(int id, int pnodecount, double* data, Datum* indexes, ThreadDatum* thread, NrnThread* nt, double v) {
        hhderiv_Instance* inst = (hhderiv_Instance*)get_memb_list(nt)->instance;
        double* savstate1 = (double*) thread[dith1()].pval;
        int* slist1 = hhderiv_global.slist1;
        int* slist2 = hhderiv_global.slist2;
        double* dlist2 = (double*) thread[dith1()].pval + (3*pnodecount);
        for (int i=0; i<3; i++) {
            savstate1[i*pnodecount+id] = data[slist1[i]*pnodecount+id];
        }
        int reset = nrn_newton_thread(static_cast<NewtonSpace*>(*newtonspace1(thread)), 3, slist2, _derivimplicit_states_hhderiv, dlist2, id, pnodecount, data, indexes, thread, nt, v);
        return reset;
    }


// somewhere in another header file _kinderiv.h
#pragma acc routine seq
int _newton_states_hhderiv(_threadargsproto_);



    /* _derivimplicit_ states _hhderiv */

    int _newton_states_hhderiv(int id, int pnodecount, double* data, Datum* indexes, ThreadDatum* thread, NrnThread* nt, double v)  {
        hhderiv_Instance* inst = (hhderiv_Instance*)get_memb_list(nt)->instance;
        double* savstate1 = (double*) thread[dith1()].pval;
        int* slist1 = hhderiv_global.slist1;
        int* dlist1 = hhderiv_global.dlist1;
        double celsius=1;
        double* dlist2 = (double*) thread[dith1()].pval + (3*pnodecount);
        {
            double alpha, beta, sum, q10, vtrap_in_0, vtrap_in_1, v_in_1;
            v_in_1 = v;
            q10 = pow(3.0, ((celsius - 6.3) / 10.0));
```

Note that `_newton_states_hhderiv` is declared as `routine seq` and it's using `static double celcisus` variable.  This result into compilation error:

```console
$ nrnivmodl -coreneuron ../mod/
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 108.25 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable vtrap at 113.17-21 to ret_vtrap
[NMODL] [warning] :: RenameVisitor :: Renaming variable vtrap at 115.17-21 to ret_vtrap
NVC++-S-1053-External and Static variables are not supported in acc routine - _ZN10coreneuron7celsiusE (x86_64/corenrn/mod2c/hhderiv.cpp: 436)
NVC++/x86-64 Linux 21.2-0: compilation completed with severe errors
make: *** [x86_64/corenrn/build/hhderiv.o] Error 2
make: *** Waiting for unfinished jobs.... 
```

## How to reproduce?

* Install neuron+coreneuron+nmodl with GPU support
* Compile mod files from testcorenrn repo

```
cmake ..  -DNRN_ENABLE_TESTS=ON -DNRN_ENABLE_CORENEURON=ON -DCORENRN_ENABLE_GPU=ON -DCORENRN_ENABLE_NMODL=ON -DCMAKE_INSTALL_PREFIX=`pwd`/install
make -j12
make install

`pwd`/install/bin/nrnivmodl-core -coreneuron mod 
```

--- a/src/codegen/codegen_acc_visitor.cpp
+++ b/src/codegen/codegen_acc_visitor.cpp
@@ -231,7 +231,7 @@ void CodegenAccVisitor::print_newtonspace_transfer_to_device() const {
     int list_num = info.derivimplicit_list_num;
     printer->add_line("if (nt->compute_gpu) {");
     printer->add_line("    auto device_vec = static_cast<double*>(acc_copyin(vec, vec_size));");
-    printer->add_line("    auto device_ns = static_cast<NewtonSpace*>(acc_deviceptr(ns));");
+    printer->add_line("    auto device_ns = static_cast<NewtonSpace*>(acc_deviceptr(*ns));");
     printer->add_line("    auto device_thread = static_cast<ThreadDatum*>(acc_deviceptr(thread));");
     printer->add_line(
         "    acc_memcpy_to_device(&(device_thread[{}]._pvoid), &device_ns, sizeof(void*));"_format(
@@ -242,6 +242,18 @@ void CodegenAccVisitor::print_newtonspace_transfer_to_device() const {
     printer->add_line("}");
 }
 
+
+void CodegenAccVisitor::print_instance_variable_transfer_to_device() const {
+    printer->add_line("if (nt->compute_gpu) {");
+    printer->add_line("    auto dml = (Memb_list*) acc_deviceptr(ml);");
+    printer->add_line("    acc_memcpy_to_device(&(dml->instance), &inst, sizeof(void*));");
+    printer->add_line("}");
+}
+
+
+void CodegenAccVisitor::print_deriv_advance_flag_transfer_to_device() const {
+    printer->add_line("#pragma acc update device (deriv_advance_flag) if (nt->compute_gpu)");
+}
 
 }  // namespace codegen
 }  // namespace nmodl

--- a/src/codegen/codegen_acc_visitor.hpp
+++ b/src/codegen/codegen_acc_visitor.hpp
@@ -88,6 +88,12 @@ class CodegenAccVisitor: public CodegenCVisitor {
     /// transfer newtonspace structure to device
     void print_newtonspace_transfer_to_device() const override;
 
+    // update instance variable object pointer on the gpu device
+    void print_instance_variable_transfer_to_device() const override;
+
+    // update derivimplicit advance flag on the gpu device
+    void print_deriv_advance_flag_transfer_to_device() const override;
+
     std::string get_variable_device_pointer(const std::string& variable,
                                             const std::string& type) const override;
 

--- a/src/codegen/codegen_c_visitor.cpp
+++ b/src/codegen/codegen_c_visitor.cpp
@@ -1030,6 +1030,13 @@ void CodegenCVisitor::print_channel_iteration_tiling_block_end() {
     // backend specific, do nothing
 }
 
+void CodegenCVisitor::print_instance_variable_transfer_to_device() const {
+    // backend specific, do nothing
+}
+
+void CodegenCVisitor::print_deriv_advance_flag_transfer_to_device() const {
+    // backend specific, do nothing
+}
 
 /**
  * \details Each kernel such as \c nrn\_init, \c nrn\_state and \c nrn\_cur could be offloaded
@@ -3169,6 +3176,7 @@ void CodegenCVisitor::print_instance_variable_setup() {
     }
 
     printer->add_line("ml->instance = (void*) inst;");
+    print_instance_variable_transfer_to_device();
     printer->end_block(3);
 
     printer->add_line("/** cleanup mechanism instance variables */");
@@ -3281,7 +3289,9 @@ void CodegenCVisitor::print_nrn_init(bool skip_init_check) {
         int nequation = info.num_equations;
         int list_num = info.derivimplicit_list_num;
         // clang-format off
-        printer->add_line("*deriv{}_advance(thread) = 0;"_format(list_num));
+        printer->add_line("int& deriv_advance_flag = *deriv{}_advance(thread);"_format(list_num));
+        printer->add_line("deriv_advance_flag = 0;");
+        print_deriv_advance_flag_transfer_to_device();
         printer->add_line("auto ns = newtonspace{}(thread);"_format(list_num));
         printer->add_line("auto& th = thread[dith{}()];"_format(list_num));
 
@@ -3313,7 +3323,8 @@ void CodegenCVisitor::print_nrn_init(bool skip_init_check) {
     printer->end_block(1);
 
     if (info.derivimplicit_used()) {
-        printer->add_line("*deriv{}_advance(thread) = 1;"_format(info.derivimplicit_list_num));
+        printer->add_line("deriv_advance_flag = 1;");
+        print_deriv_advance_flag_transfer_to_device();
     }
     print_kernel_data_present_annotation_block_end();
     if (skip_init_check) {

--- a/src/codegen/codegen_c_visitor.hpp
+++ b/src/codegen/codegen_c_visitor.hpp
@@ -1082,6 +1082,17 @@ class CodegenCVisitor: public visitor::ConstAstVisitor {
 
 
     /**
+     * Print the code to copy instance variable to device
+     */
+    virtual void print_instance_variable_transfer_to_device() const;
+
+    /**
+     * Print the code to copy derivative advance flag to device
+     */
+    virtual void print_deriv_advance_flag_transfer_to_device() const;
+
+
+    /**
      * Print byte arrays that register scalar and vector variables for hoc interface
      *
      */

